lxd/images.go

```diff
@@ -77,183 +84,271 @@ type imageMetadata struct {
+	if err = decoder.Decode(&req); err == nil {
+		return imgPostContInfo(r, req, builddir)
 	}
 
-	public, err := strconv.Atoi(r.Header.Get("X-LXD-public"))
-	tarname := r.Header.Get("X-LXD-filename")
+	// ok we've got an image in the body
+	public, _ = strconv.Atoi(r.Header.Get("X-LXD-public"))
+	filename = r.Header.Get("X-LXD-filename")
+	propHeaders := r.Header[http.CanonicalHeaderKey("X-LXD-properties")]
 
-	dirname := shared.VarPath("images")
-	err = os.MkdirAll(dirname, 0700)
-	if err != nil {
-		return InternalError(err)
+	properties = map[string]string{}
+	if len(propHeaders) > 0 {
+		for _, ph := range propHeaders {
+			p, _ := url.ParseQuery(ph)
+			for pkey, pval := range p {
+				properties[pkey] = pval[0]
+			}
+		}
 	}
 
-	f, err := ioutil.TempFile(dirname, "lxd_image_")
+	// Create a file for the tarball
+	tarf, err := ioutil.TempFile(builddir, "lxd_tar_")
 	if err != nil {
-		return InternalError(err)
+		return 0, "", 0, "", 0, properties, err
 	}
-	defer f.Close()
-
-	fname := f.Name()
 
+	tarfname := tarf.Name()
 	sha256 := sha256.New()
-	size, err := io.Copy(io.MultiWriter(f, sha256), r.Body)
+
+	var size1, size2 int64
+	size1, err = io.Copy(io.MultiWriter(tarf, sha256), decoder.Buffered())
+	if err == nil {
+		size2, err = io.Copy(io.MultiWriter(tarf, sha256), r.Body)
+	}
+	size = size1 + size2
+	tarf.Close()
 	if err != nil {
-		return cleanup(err, fname)
+		return 0, "", 0, "", 0, properties, err
 	}
 
-	fingerprint := fmt.Sprintf("%x", sha256.Sum(nil))
+	fingerprint = fmt.Sprintf("%x", sha256.Sum(nil))
 	expectedFingerprint := r.Header.Get("X-LXD-fingerprint")
 	if expectedFingerprint != "" && fingerprint != expectedFingerprint {
 		err = fmt.Errorf("fingerprints don't match, got %s expected %s", fingerprint, expectedFingerprint)
-		return cleanup(err, fname)
+		return 0, "", 0, "", 0, properties, err
 	}
 
-	imagefname := shared.VarPath("images", fingerprint)
-	if shared.PathExists(imagefname) {
-		return cleanup(fmt.Errorf("Image already exists."), fname)
+	imagefname := fmt.Sprintf("%s/%s", builddir, fingerprint)
+	err = os.Rename(tarfname, imagefname)
+	if err != nil {
+		return 0, "", 0, "", 0, properties, err
 	}
 
-	err = os.Rename(fname, imagefname)
+	var imageMeta *imageMetadata
+	imageMeta, err = getImageMetadata(imagefname)
 	if err != nil {
-		return cleanup(err, fname)
+		return 0, "", 0, "", 0, properties, err
 	}
 
-	if backing_fs == "btrfs" {
-		subvol := fmt.Sprintf("%s.btrfs", imagefname)
-		output, err := exec.Command("btrfs", "subvolume", "create", subvol).CombinedOutput()
-		if err != nil {
-			shared.Debugf("btrfs subvolume creation failed\n")
-			shared.Debugf(string(output))
-			return cleanup(err, imagefname)
-		}
+	arch, _ = shared.ArchitectureId(imageMeta.Architecture)
 
-		compression, _, err := detectCompression(imagefname)
-		if err != nil {
-			return cleanup(err, imagefname)
-		}
+	err = nil
+	return
+}
+
+func makeBtrfsFile(imagefname, subvol string) error {
```

Comment:
I was a little confused by the name 'makeBtrfsFile', should it be something like 'makeBtrfsSubvolume'?